nessie/jobs/sync_canvas_requests_snapshots.py

```diff
@@ -0,0 +1,111 @@
+INCIDENTAL, OR CONSEQUENTIAL DAMAGES, INCLUDING LOST PROFITS, ARISING OUT OF
+THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN IF REGENTS HAS BEEN ADVISED
+OF THE POSSIBILITY OF SUCH DAMAGE.
+
+REGENTS SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
+SOFTWARE AND ACCOMPANYING DOCUMENTATION, IF ANY, PROVIDED HEREUNDER IS PROVIDED
+"AS IS". REGENTS HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES,
+ENHANCEMENTS, OR MODIFICATIONS.
+"""
+
+import time
+
+from flask import current_app as app
+from nessie.externals import canvas_data, s3
+from nessie.jobs.background_job import BackgroundJob, BackgroundJobError
+from nessie.lib import metadata
+from nessie.lib.dispatcher import dispatch
+
+"""Logic for Canvas snapshot sync job."""
+
+
+def delete_objects_with_prefix(prefix, whitelist=[]):
+    keys_to_delete = []
+    existing_keys = s3.get_keys_with_prefix(prefix)
+    if existing_keys is None:
+        app.logger.error('Error listing keys, aborting job.')
+        return False
+    for key in existing_keys:
+        filename = key.split('/')[-1]
+        if filename not in whitelist:
+            keys_to_delete.append(key)
+    app.logger.info(
+        f'Found {len(existing_keys)} key(s) matching prefix "{prefix}", {len(existing_keys) - len(keys_to_delete)} '
+        f'key(s) in whitelist, will delete {len(keys_to_delete)} object(s)')
+    if not keys_to_delete:
+        return True
+    if s3.delete_objects(keys_to_delete):
+        metadata.delete_canvas_snapshots(keys_to_delete)
+        return True
+    else:
+        return False
+
+
+class SyncCanvasRequestsSnapshots(BackgroundJob):
+
+    @classmethod
+    def generate_job_id(cls):
+        return 'sync_canvas_requests_' + str(int(time.time()))
+
+    def run(self, cleanup=True):
+        job_id = self.generate_job_id()
+        app.logger.info(f'Starting Canvas snapshot sync job... (id={job_id})')
+
+        snapshot_response = canvas_data.get_snapshots()
+        if not snapshot_response:
+            raise BackgroundJobError('Error retrieving Canvas data snapshots, aborting job.')
+        snapshots = snapshot_response.get('files', [])
+
+        def should_sync(snapshot):
+            return snapshot['table'] == 'requests' and snapshot['partial'] is False
+
+        snapshots_to_sync = [s for s in snapshots if should_sync(s)]
+        app.logger.info(f'Will sync {len(snapshots_to_sync)} of {len(snapshots)} available files from Canvas Data.')
+
+        success = 0
+        failure = 0
+
+        for snapshot in snapshots_to_sync:
+            metadata.create_canvas_sync_status(
+                job_id=job_id,
+                filename=snapshot['filename'],
+                canvas_table=snapshot['table'],
+                source_url=snapshot['url'],
+            )
+            if snapshot['table'] == 'requests':
```

Comment:
@sandeepmjay This conditional shouldn't be necessary if `snapshot['table']` will always be `requests`?